crates/transaction-pool/src/batcher.rs

```diff
@@ -0,0 +1,273 @@
+    BatcherChannelClosed,
+    /// Transaction response channel closed
+    #[error("Transaction response channel closed")]
+    ResponseChannelClosed,
+}
+
+/// A single batch transaction request
+/// All transactions processed through the batcher are considered local
+/// transactions (`TransactionOrigin::Local`) when inserted into the pool.
+#[derive(Debug)]
+pub struct BatchTxRequest<T: PoolTransaction> {
+    /// Tx to be inserted in to the pool
+    pool_tx: T,
+    /// Channel to send result back to caller
+    response_tx: oneshot::Sender<Result<B256, BatchTxError>>,
+}
+
+impl<T> BatchTxRequest<T>
+where
+    T: PoolTransaction,
+{
+    /// Create a new batch transaction request
+    pub const fn new(pool_tx: T, response_tx: oneshot::Sender<Result<B256, BatchTxError>>) -> Self {
+        Self { pool_tx, response_tx }
+    }
+}
+
+/// Transaction batch processor that handles batch processing
+#[pin_project]
+#[derive(Debug)]
+pub struct BatchTxProcessor<Pool: TransactionPool> {
+    pool: Pool,
+    max_batch_size: usize,
+    channel_buffer_size: usize,
+    #[pin]
+    request_rx: mpsc::Receiver<BatchTxRequest<Pool::Transaction>>,
+}
+
+impl<Pool> BatchTxProcessor<Pool>
+where
+    Pool: TransactionPool + Clone + Send + Sync + 'static,
+    Pool::Transaction: Send + Sync,
+{
+    /// Create a new `BatchTxProcessor`
+    pub fn new(
+        pool: Pool,
+        max_batch_size: usize,
+        channel_buffer_size: usize,
+    ) -> (Self, mpsc::Sender<BatchTxRequest<Pool::Transaction>>) {
+        let (request_tx, request_rx) = mpsc::channel(channel_buffer_size);
+
+        let processor = Self { pool, max_batch_size, channel_buffer_size, request_rx };
+
+        (processor, request_tx)
+    }
+
+    /// Process a batch of transaction requests, grouped by origin
+    async fn process_batch(pool: &Pool, batch: Vec<BatchTxRequest<Pool::Transaction>>) {
+        let (pool_transactions, response_tx): (Vec<_>, Vec<_>) =
+            batch.into_iter().map(|req| (req.pool_tx, req.response_tx)).unzip();
+
+        let pool_results = pool.add_transactions(TransactionOrigin::Local, pool_transactions).await;
+
+        for (response_tx, pool_result) in response_tx.into_iter().zip(pool_results) {
+            let final_result = match pool_result {
+                Ok(AddedTransactionOutcome { hash, .. }) => Ok(hash),
+                Err(_) => Err(BatchTxError::BatcherChannelClosed),
+            };
+
+            let _ = response_tx.send(final_result);
+        }
+    }
+}
+
+impl<Pool> Future for BatchTxProcessor<Pool>
+where
+    Pool: TransactionPool + Clone + Send + Sync + 'static,
+    Pool::Transaction: Send + Sync,
+{
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+
+        loop {
+            // Drain all available requests from the receiver
+            let mut batch = Vec::with_capacity(*this.channel_buffer_size);
```

Comment:
I think we can get away with allocating only a few slots here maybe just two, 1 should also be fine